core/src/test/scala/unit/kafka/network/SocketServerTest.scala

```diff
@@ -593,4 +596,278 @@ class SocketServerTest extends JUnitSuite {
+        config.connectionsMaxIdleMs, listenerName, protocol, config, metrics, credentialProvider, memoryPool) {
+        override protected[network] def createSelector(maxReceiveSize: Int, connectionMaxIdleMs: Long,
+             metrics: Metrics, time: Time, metricGrpPrefix: String, metricTags: java.util.Map[String, String],
+             metricsPerConnection: Boolean, recordTimePerConnection: Boolean,
+             channelBuilder: ChannelBuilder, memoryPool: MemoryPool): Selector = {
+
+           val testableSelector = new TestableSelector(maxReceiveSize, connectionMaxIdleMs, metrics, time, metricGrpPrefix, metricTags,
+                            metricsPerConnection, recordTimePerConnection, channelBuilder, memoryPool)
+           selector = Some(testableSelector)
+           testableSelector
+        }
+      }
+    }
+    def testableSelector = selector.getOrElse(throw new IllegalStateException("Selector not created"))
+
+    def waitForChannelException(connectionId: String, socket: Socket, operation: SelectorOperation.Value): Unit = {
+      val selector = testableSelector
+      selector.waitForOperations(operation, 1)
+      TestUtils.waitUntilTrue(() =>  selector.allLocallyClosedChannels.contains(connectionId), "Channel not closed")
+      assertEquals(0, connectionCount(socket.getLocalAddress))
+      assertTrue("Inflight responses not cleared", processor(0).inflightResponses.isEmpty)
+      assertNull("Channel not removed", selector.channel(connectionId))
+      assertNull("Closing channel not removed", selector.closingChannel(connectionId))
+      assertTrue("Unexpected disconnect notification",testableSelector.allDisconnectedChannels.isEmpty)
+    }
+  }
+
+  object SelectorOperation extends Enumeration {
+    type SelectorOperation = Value
+    val REGISTER, POLL, SEND, MUTE, UNMUTE, CLOSE, DISCONNECTED = Value
+  }
+
+  class TestableSelector(maxReceiveSize: Int, connectionMaxIdleMs: Long,
+                         metrics: Metrics, time: Time, metricGrpPrefix: String, metricTags: java.util.Map[String, String], metricsPerConnection: Boolean,
+                         recordTimePerConnection: Boolean, channelBuilder: ChannelBuilder, memoryPool: MemoryPool)
+        extends Selector(maxReceiveSize, connectionMaxIdleMs, metrics, time, metricGrpPrefix, metricTags, metricsPerConnection,
+                         recordTimePerConnection, channelBuilder, memoryPool) {
+
+    val failures = new mutable.HashMap[SelectorOperation.Value, Exception]
+    val operationCounts = new mutable.HashMap[SelectorOperation.Value, Int]
+    val allChannels = new mutable.HashSet[String]
+    val allLocallyClosedChannels = new mutable.HashSet[String]
+    val allDisconnectedChannels = new mutable.HashSet[String]
+    @volatile var closed = false
+
+    private def onOperation(operation: SelectorOperation.Value): Unit = {
+      operationCounts.put(operation, operationCounts.getOrElse(operation, 0) + 1)
```

Comment:
You can use `getOrElseUpdate`.

Reply:
Want to update in both cases, so added `withDefaultValue`.